flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/ProjectMultiJoinTransposeRule.java

```diff
@@ -0,0 +1,491 @@
+
+        final Map<Integer, ImmutableIntList> newJoinFieldRefCountsMap = new HashMap<>();
+        final Map<Integer, ImmutableIntList> originalJoinFieldRefCountsMap =
+                originalMultiJoin.getJoinFieldRefCountsMap();
+
+        for (Map.Entry<Integer, ImmutableIntList> entry :
+                originalJoinFieldRefCountsMap.entrySet()) {
+            final Integer inputIndex = entry.getKey();
+            final ImmutableIntList originalRefCounts = entry.getValue();
+            final RelNode newInput = transformedInputs.newInputs.get(inputIndex);
+            final int newFieldCount = newInput.getRowType().getFieldCount();
+
+            // Create new ref counts array for this input's new field count
+            final int[] newRefCounts = new int[newFieldCount];
+
+            // Map the original field indices to new field indices for this input
+            for (int originalFieldIndex = 0;
+                    originalFieldIndex < originalRefCounts.size();
+                    originalFieldIndex++) {
+                // Calculate the global field index for this input's field
+                final int globalFieldIndex =
+                        calculateGlobalFieldIndex(
+                                inputIndex, originalFieldIndex, originalMultiJoin.getInputs());
+                final int newGlobalFieldIndex = fieldMapping.getTargetOpt(globalFieldIndex);
+
+                if (newGlobalFieldIndex != -1) {
+                    // Calculate the new local field index within this input
+                    final int newLocalFieldIndex =
+                            calculateLocalFieldIndex(
+                                    newGlobalFieldIndex, transformedInputs.newInputs, inputIndex);
+                    if (newLocalFieldIndex >= 0 && newLocalFieldIndex < newFieldCount) {
+                        newRefCounts[newLocalFieldIndex] =
+                                originalRefCounts.get(originalFieldIndex);
+                    }
+                }
+            }
+
+            newJoinFieldRefCountsMap.put(inputIndex, ImmutableIntList.of(newRefCounts));
+        }
+
+        return newJoinFieldRefCountsMap;
+    }
+
+    /** Safely checks if a RelNode is a Project, handling both HepRelVertex and RelSubset. */
+    private boolean isProject(RelNode relNode) {
+        if (relNode instanceof Project) {
+            return true;
+        }
+
+        // Handle HepRelVertex (HEP planner)
+        if (relNode instanceof HepRelVertex) {
+            return ((HepRelVertex) relNode).getCurrentRel() instanceof Project;
+        }
+
+        return false;
+    }
+
+    /** Calculates the global field index for a field within a specific input. */
+    private int calculateGlobalFieldIndex(
+            int inputIndex, int localFieldIndex, List<RelNode> inputs) {
+        int globalFieldIndex = 0;
+        for (int i = 0; i < inputIndex; i++) {
+            globalFieldIndex += inputs.get(i).getRowType().getFieldCount();
+        }
```

Comment:
this part seems to be duplicated, can we extract it?

Reply:
I've refactored it, take a look